server/rest.go

```diff
@@ -1313,55 +1330,44 @@ func (s *Server) UpdateContract(c echo.Context) error {
 	)
 	c.Request().Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
 	if err := c.Bind(&contract); err != nil {
-		lgr.Error("cannot bind data", zap.Error(err))
+		lgr.Error("cannot bind contract data", zap.Error(err))
 		return api.Invalid.Build(c)
 	}
 	c.Request().Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
 	if err := c.Bind(&addrInfo); err != nil {
-		lgr.Error("cannot bind data", zap.Error(err))
+		lgr.Error("cannot bind address data", zap.Error(err))
 		return api.Invalid.Build(c)
 	}
 	ctx := context.Background()
-	if strings.EqualFold(addrInfo.ErcTypes, "KRC20") {
-		smcABIStr, err := s.dbClient.SMCABIByType(ctx, addrInfo.ErcTypes)
-		if err == nil {
-			abiData, err := base64.StdEncoding.DecodeString(smcABIStr)
-			if err != nil {
-				s.logger.Warn("Cannot decode smc abi", zap.Error(err))
-			}
-			jsonABI, err := abi.JSON(bytes.NewReader(abiData))
-			if err != nil {
-				s.logger.Warn("Cannot convert decoded smc abi to JSON abi", zap.Error(err))
-			}
-			if err == nil {
-				totalSupply, _ := s.kaiClient.GetKRCTotalSupply(ctx, &jsonABI, common.HexToAddress(addrInfo.Address))
-				if totalSupply != nil {
-					addrInfo.TotalSupply = totalSupply.String()
-				}
-			}
-		}
+	krcTokenInfoFromRPC, err := s.getKRCTokenInfoFromRPC(ctx, addrInfo.Address, addrInfo.KrcTypes)
+	if err != nil && strings.HasPrefix(addrInfo.KrcTypes, "KRC") {
+		s.logger.Warn("Updating contract is not KRC type", zap.Any("smcInfo", addrInfo), zap.Error(err))
+		return api.Invalid.Build(c)
+	}
+	if krcTokenInfoFromRPC != nil {
+		// cache new token info
+		krcTokenInfoFromRPC.Logo = addrInfo.Logo
+		_ = s.cacheClient.UpdateKRCTokenInfo(ctx, krcTokenInfoFromRPC)
+
+		addrInfo.TokenName = krcTokenInfoFromRPC.TokenName
+		addrInfo.TokenSymbol = krcTokenInfoFromRPC.TokenSymbol
+		addrInfo.TotalSupply = krcTokenInfoFromRPC.TotalSupply
+		addrInfo.Decimals = krcTokenInfoFromRPC.Decimals
+		addrInfo.NumTokensTotal = krcTokenInfoFromRPC.NumTokensTotal
 	}
 	currTokenInfo, _ := s.dbClient.AddressByHash(ctx, addrInfo.Address)
 	if err := s.dbClient.UpdateContract(ctx, &contract, &addrInfo); err != nil {
 		lgr.Error("cannot bind insert", zap.Error(err))
 		return api.InternalServer.Build(c)
 	}
```

Comment:
nil currToken is ok ?

Reply:
It's ok, I will checked it later at `if (currTokenInfo != nil && currTokenInfo.KrcTypes == "" && currTokenInfo.TokenName == "" && currTokenInfo.TokenSymbol == "") || currTokenInfo == nil {`